ocaml/xapi/xapi_cluster.ml

```diff
@@ -100,17 +108,79 @@ let pool_create ~__context ~network ~cluster_stack ~token_timeout ~token_timeout
 
   cluster
 
+(* Helper function; if opn is None return all, else return those not equal to it *)
+let filter_on_option opn xs =
+  match opn with
+  | None -> xs
+  | Some x -> List.filter ((<>) x) xs
+
+(* Helper function; concurrency checks are done in implementation of Cluster.destroy and Cluster_host.destroy *)
+let pool_force_destroy ~__context ~self = 
+  (* For now we assume we have only one pool, and that the cluster is the same as the pool.
+     This means that the pool master must be a member of this cluster. *)
+  let master = Helpers.get_master ~__context in
+  let master_cluster_host =
+    Xapi_clustering.find_cluster_host ~__context ~host:master
+  in
+  let slave_cluster_hosts =
+    Db.Cluster.get_cluster_hosts ~__context ~self |> filter_on_option master_cluster_host
+  in
+  (* First try to destroy each cluster_host - if we can do so safely then do *)
+  List.iter
+    (fun cluster_host -> 
+      (* We need to run this code on the slave *)
+      (* We ignore failures here, we'll try a force_destroy after *)
+      log_and_ignore_exn (fun () ->  
+        Helpers.call_api_functions ~__context (fun rpc session_id ->
+            Client.Client.Cluster_host.destroy ~rpc ~session_id ~self:cluster_host)
+      )
+    )
+    slave_cluster_hosts;
+  (* We expect destroy to have failed for some, we'll try to force destroy those *)
+  (* Note we include the master here, we should attempt to force destroy it *)
+  let all_remaining_cluster_hosts =
+    Db.Cluster.get_cluster_hosts ~__context ~self
+  in
+  (* Now try to force_destroy, keep track of any errors here *)
+  let exns = List.fold_left
+    (fun exns_so_far cluster_host ->
+      Helpers.call_api_functions ~__context (fun rpc session_id ->
+        try
+          Client.Client.Cluster_host.force_destroy ~rpc ~session_id ~self:cluster_host;
+          exns_so_far
+        with e ->
+          Backtrace.is_important e;
+          let uuid = Client.Client.Cluster_host.get_uuid ~rpc ~session_id ~self:cluster_host in
+          debug "Ignoring exception while trying to force destroy cluster host %s: %s" uuid (ExnHelper.string_of_exn e);
+          exns_so_far @ [e]
```

Comment:
It is usually better to construct lists by adding value to head (`O(1)`) vs concatenating (`O(n)`). You will get the list in reverse order, but thats fine in this case.